libs/metrics/in_out.py

```diff
@@ -1,25 +1,58 @@
-import os
+import ast
+import csv
 import json
+import os
 import numpy as np
+
+from numpy import linalg as LA
+from collections import deque
 from datetime import date
 from typing import Dict, Iterator, List
+from datetime import datetime
 
 from .base import BaseMetric
 from constants import IN_OUT
 from libs.utils.config import get_source_config_directory
 from libs.utils.utils import validate_file_exists_and_is_not_empty
-from pathlib import Path
 
 class InOutMetric(BaseMetric):
 
-    reports_folder = IN_OUT
-    csv_headers = []
-    entity = "area"
-    live_csv_headers = []
+    metric_name = IN_OUT
+    reports_folder = metric_name
+    csv_headers = ["In", "Out"]
 
     @classmethod
-    def process_csv_row(cls, csv_row: Dict, objects_logs: Dict):
-        raise NotImplementedError("Operation not implemented")
+    def process_path(cls, boundary_line, trajectory_path):
+        """
+        Verify if a trajectory goes over a boundary line
+        Args:
+            Two coordinates [x,y] are in 2-tuples [A,B]
+                Boundaries of the in/out line.
+                If someone crosses the line while having A to their right, they are going in the in direction (entering)
+                Crossing the line while having A to their left means they are going in the out direction (leaving)
+
+            trajectory_path: List of N 2-tuples (x,y)
+            That represents the trajectory of an object.
+
+        Returns:
+            in, out) : tuple
+                 (1, 0) - if the object entered (in)
+                 (0, 1) - if the object left (out)
+                 (0, 0) - if the object didn't cross the boundary.
+        """
+        number_of_cuts = 5
+        if len(trajectory_path) < number_of_cuts:
+            number_of_cuts = len(trajectory_path)
+
+        intersections = [trajectory_path[int(i)] for i in np.linspace(0, len(trajectory_path) - 1, number_of_cuts)]
+        intersections = zip(intersections, intersections[1:])
+        in_out = (0, 0)
+
+        for intersection in intersections:
```

Comment:
This will return the last crossing of the line only. I think it is good to return only one value, unless the camera covers a big space and a person could be in the camera's space for some time and go out. In that case we need to count 1 in and 1 out.
Also if a person comes from outside and shortly goes over the line and back you will count this a 0 in and 1 out with this code.

Ideally I think we would mix the time dimension in and say that the person needs to be "inside" for X seconds before going out to be counted if it goes in and out in the same "session". Same applies in the opposite direction

Reply:
So you'd rather count every intersection with the line and return the sum of in_out rather than having policyy for selecting one?

Reply:
No. We should only return 1. Or maybe 1 for in and 1 for out (this if we check the time a user spents in or out, likely not v1)

But with the current code if a user goes in, and before leaving the screen goes out you are counting 0 in and 1 out which is not correct

Reply:
Ah, I see what you mean. So then we should return (1,1), right?

My issue there is that if the user goes in, out, and then in again it should be (1,0) or (2,1).

Maybe we could return (1,0) by normalizing (2,1) to (1,0), would  that make  sense?

Reply:
I think that the correctness of changing (2,1) to (1,0) depends on the time the person spends in and out. Like if the bbox detection flickers a bit and jumps above and below the line then we don't want to count that as multiple crosses. But if the user spends some time then maybe yes. We could also define this as a minimum of N frames on one side before we count a user that goes back and forth as (1,1). If it is les than N frames we take it as (0,0)